src/daft-logical-plan/src/optimization/rules/extract_window_function.rs

```diff
@@ -0,0 +1,488 @@
+use std::sync::Arc;
+
+use common_error::DaftResult;
+use common_treenode::{Transformed, TreeNode};
+use daft_dsl::{expr::window::WindowSpec, resolved_col, Expr, ExprRef};
+use itertools::Itertools;
+
+use crate::{
+    logical_plan::{LogicalPlan, Project},
+    ops::Window,
+    optimization::rules::OptimizerRule,
+};
+
+#[derive(Debug)]
+pub struct ExtractWindowFunction;
+
+impl Default for ExtractWindowFunction {
+    fn default() -> Self {
+        Self
+    }
+}
+
+impl ExtractWindowFunction {
+    pub fn new() -> Self {
+        Self
+    }
+
+    fn is_window_function_expr(expr: &ExprRef) -> bool {
+        expr.exists(|e| matches!(e.as_ref(), Expr::Window(_, _)))
+    }
+
+    fn contains_window_function(project: &Project) -> bool {
+        project.projection.iter().any(Self::is_window_function_expr)
+    }
+
+    fn extract_window_functions(projection: &[ExprRef]) -> Vec<(ExprRef, WindowSpec)> {
+        let mut result = Vec::new();
+
+        for expr in projection {
+            Self::collect_window_functions(expr, &mut result);
+        }
+
+        result
+    }
+
+    fn collect_window_functions(expr: &ExprRef, result: &mut Vec<(ExprRef, WindowSpec)>) {
+        match expr.as_ref() {
+            Expr::Window(_inner_expr, window_spec) => {
+                result.push((expr.clone(), window_spec.clone()));
+            }
+            _ => {
+                for child in expr.children() {
+                    Self::collect_window_functions(&child, result);
+                }
+            }
+        }
```

Comment:
You could use expr.apply(..) here so you do not need to do your own recursion. Once you hit a window you add it to result and then do TreeNodeRecursion::Jump